Weboob integration

Phyks opened the pull request:
Weboob integration was recently discussed in #496 

I came up with a [wrapper script](https://github.com/Phyks/cozyweboob) around weboob to ease its integration in Konnector, and here is the associated connector.

For now it is very basic, but was proven to be somehow working using my Freemobile account. I should still work on it and stabilize it, hence the "[WIP]" tag, to emphasize it should not be merged at the moment.

I open this PR to have feedbacks on the code, and discuss about some issues I had while writing it.

Of course, commits will be squashed and reworked when merging will be considered.

If you want to give it a try yourself, you should:
1. Install Weboob on your system.
2. Clone [this repo](https://github.com/Phyks/cozyweboob) somewhere and launch the `server.py` script.
3. Use the branch of this PR for Konnectors, and use dev tools to `--init` and set the required fields. Typically, I use `coffee dev_tools.coffee --change weboob --fields "{\"weboobURL\": \"http://localhost:8080\", \"JSONModulesDescription\": $(cat ../cozyweboob/konnectors.json)}"`.
4. Run an import and see how it goes.

Note that for the moment, it only handles `Bill` model, but extending it to other models is just a matter of a few LoC and time, no major troubles here.

I provide extra discussion and questions about these changes in the discussion of this PR.

Commit:
Fix scroll on mobile (#543)

JbIPS committed:
Adds a Meetup konnector (#535)

Adds a Meetup konnector. Fixes #461

ZeHiro committed:
Add logout to meetup konnector, so that multiple import can happen

Phyks committed:
Weboob konnector
as recently discussed in #496

[Cozyweboob](https://github.com/Phyks/cozyweboob) is a wrapper script
around [weboob](http://weboob.org/) to ease its integration in
Konnectors, and here is the associated connector.

If you want to use it, you should:
* Install [Weboob](http://weboob.org/) on your system.
* Clone [cozyweboob](https://github.com/Phyks/cozyweboob) somewhere and
launch the `server.py` script.
* Run an import using the `weboob` connector..

_Note_: for the moment, it only handles `Bill` model, but extending it to
other models is straightforward.

Fixes #496.

* Add missing fields in `Bill`, compared to Weboob output:
    - `vat` to store the VAT part of the total amount.
    - `currency` to support bills in different currencies.
    - Extra date fields (`startdate`, `finishdate`, `duedate` )from weboob.
* Add some doc on `Bill` model, especially fields description as
comments.

Phyks commented on dev_tools.coffee:
```diff
@@ -34,8 +34,8 @@ else if program.change
         konnectorName = program.change
-        fields = program.fields.split(',')
-        log.info "Changing fields for #{konnectorName} with #{fields}."
+        fields = JSON.parse(program.fields)
```
Major change in dev tools, I require the `fields` attribute to be a JSON string describing the fields to set. This was the easiest way to get the dev tools to let me set the weboob konnectors required fields. But that may not be something we want to merge as is.

Phyks commented on server/konnectors/weboob.js:
```diff
@@ -0,0 +1,206 @@
+ * You need a running [cozyweboob](https://github.com/Phyks/cozyweboob)
+ * somewhere.
+ *
+ * // TODO: What happens when moment is an invalid date?
```
Seems to be a major trouble actually. What is a "safe" value for "no date"?

nicofrand replied:
In cases where the date for a given bill is not returned ?
Maybe now ?

Phyks replied:
Yes, typically if whatever module did not manage to retrieve a date for the bill. A bill will almost always have a date, and this should not be an issue, but it may not have duedate / startdate / finishdate. Should these just be set to "null" ? "undefined" ?

Phyks replied:
Used null. I guess it should be ok looking at the code.

Phyks commented on server/konnectors/weboob.js:
```diff
@@ -0,0 +1,206 @@
+    };
+    entries.fetched = data.parsedEntries[Bill];
+    if (entries.fetched !== undefined) {
+        saveDataAndFile(weboobKonnector.logger, Bill, fileOptions, ['bill']) (requiredFields, entries, data, next);
```
Same again, not sure what is expected to be put instead of `['bill']`.

Phyks commented on server/models/bill.coffee:
```diff
@@ -2,17 +2,22 @@ cozydb = require 'cozydb'
 
 
 module.exports = Bill = cozydb.getModel 'Bill',
```
I tried to comment what each field should contain. Not sure I guessed all of them correctly though.

Phyks commented on server/models/bill.coffee:
```diff
@@ -2,17 +2,22 @@ cozydb = require 'cozydb'
+    amount: Number  # Bill amount
+    vat: Number  # VAT part on the bill
+    currency: String  # Currency of the bill
+    plan: String  # TODO
```
Only use case is in [https://github.com/cozy-labs/konnectors/blob/23c8e3ef4354ef94e876e296695702df3a6a5fbc/server/konnectors/github.coffee](Github connector). Not sure what it should be set to, and looks like it could be a duplicate with other fields.

ZeHiro commented on server/models/bill.coffee:
```diff
@@ -2,17 +2,22 @@ cozydb = require 'cozydb'
+    plan: String  # TODO
+    pdfurl: String  # Link to the bill document (not necessarily a PDF file)
+    binaryId: String  # TODO
+    fileId: String  # TODO
```
It is the id of the file doctype containing the bill. The bill is just a file + metadata. Binaryid it the id of the file content

Phyks replied:
Ok, I added comments for this.

Phyks commented:
- [ ] Also, for now it has only been tested from the CLI dev tools. Web interfaces is not the most user friendly possible, as the user would have an input field to put some long JSON string describing all the modules from weboob he wants to use. [cozyweboob](https://github.com/Phyks/cozyweboob) has [an endpoint](https://github.com/Phyks/cozyweboob/blob/master/server.py#L67) to list all available modules and config options, and I think we should use it to build a nice form page to enable modules and configure them in the Konnector options. This part is not straightforward at the moment (at least, I am not sure how to handle it).

Phyks commented on server/konnectors/weboob.js:
```diff
@@ -0,0 +1,206 @@
+                    vat: parseFloat(bill.vat),
+                    currency: bill.currency,
+                    plan: '',  // TODO: What is it?
+                    pdfurl: bill.url,
```
There is a bug there at the moment, as we always store the URL, even if we [only asked for a fetch](https://github.com/Phyks/cozyweboob#input-json-file). In the `fetch` only case, the URL is set to the original URL on the vendor site, and we basically end up downloading a login page.

Phyks replied:
Done

Phyks commented on server/konnectors/weboob.js:
```diff
@@ -0,0 +1,206 @@
+        JSON.parse(requiredFields.JSONModulesDescription),
+        function (err, res, fetchedItems) {
+            // Store fetched entries
+            console.assert(res.statusCode == 200);
```
TODO: Better error handling across this connector :)

Phyks replied:
Done

ZeHiro committed:
Warn the user that the password cannot be decrypted. Fixes #503 (#538)

Phyks committed:
Fix `filter_existing`

Allow to use `null` as a safe default value for a missing date. For the
moment, simply discard such entries to keep a consistent behavior with
what was existing before.

Phyks committed:
Use conversational script for weboob konnector

Phyks commented on server/konnectors/weboob.js:
```diff
@@ -0,0 +1,204 @@
+ */
+function fetchData(requiredFields, entries, data, next) {
+    var client = requestJson.createClient(requiredFields.weboobURL);
+    client.post(
```
Done

Phyks committed:
Handle download properly, as well as cleaning before exiting

Phyks committed:
Fix File model to support creation from local files and not only URLs

Phyks commented:
So, this should work almost as expected by now, but I would need some help to finalize it. So far, I can import Freemobile bills (quite) well using CLI dev tools and `cozy-dev` VM.

## Implemented features

* Weboob wrapper, supporting translation from Weboob bills to Cozy Bills model.
* Refactor of dev tools (see note below).
* Fix `filterExisting` to allow one to use `null` as a default safe value for a missing date. If such a date is `null`, the behavior is simply to filter out the entry, to keep the actual behavior.
* Fix `File` model to allow one to create a `File` based on a local file rather than a remote URL (as `cozyweboob` download bills locally).

## How to test it

So, first, for anyone interested in giving it a try, here is a step by step guide:
* First, clone my branch of this PR.
* Then, do a `npm install` (as I added a dependency on `python-shell`) and a `git submodule init/update` (as `cozyweboob` is fetched as a submodule).
* Do not forget to install `weboob` globally so that the `cozyweboob` Python script can import it.
* Then, build the server and client using `npm run build`.
* Then, using the dev tools, you can `npm run dev_tools -- --init` to initialize konnectors and set fields. Typically, I do a `npm run dev_tools -- --change weboob --fields '{"folderPath": "/", "JSONModulesDescription": "$(cat konnectors.json)"}'` where `konnectors.json` contains:

```
[
    {
        "id": "freemobile",
        "name": "freemobile",
        "parameters": {
            "login": MY_ID",
            "password": "MY_PASSWORD"
        },
        "actions": {
            "fetch": true,
            "download": true
        }
    }
]
```

* Then, `npm run dev_tools -- --import weboob` to start an import.

It should fetch the bills from your Freemobile account and put them in your Cozy instance (or `cozy-dev` VM).


## Now, what is not working

* This only supports `Bill` at the moment (and I could only test it with Freemobile for now), but extending it further should not be really difficult.

* I have an issue with dates. There is sometimes an offset of one day that appears out of nowhere. This means that `filterExisting` is unable to match the bills (typically there is an offset of one day between the hash and the date in the object). This also means that [saveDataAndFile](https://github.com/cozy-labs/konnectors/blob/master/server/lib/save_data_and_file.coffee#L93) duplicates **all** the files. I have no idea where this is coming from... :/

* There are some `TODO`s in the code, mainly for fields which I did not know how to map, or what they should contain. That could be fixed easily as well.

* Downloaded files do not show up in the "Files" app ("Files" tells me "This folder is empty"), although everything seems to be fine in the data browser.
<img width="1439" alt="capture d ecran 2016-11-01 a 17 28 41" src="https://cloud.githubusercontent.com/assets/3856586/19908250/c6e0497e-a058-11e6-8fc6-2d76fc872007.png">


_Note_: My other PR, about rewriting and enhancing the dev tools is included in this one, as `weboob` module is not usable without it. It should be merged (or refused) before Weboob, and I will rebase and squash this PR afterwards. I will reinclude build files before merging, if needed, but removed them at the moment to ease readability.

cc @m4dz @ZeHiro @nicofrand

Phyks committed:
Add description locales for weboob connector

Phyks committed:
Enhance birthdays konnector (#565)

* Fix typos in description and update it according to latest changes.
* Make the `tag` field optional, defaulting to selecting all contacts
when the `tag` field is empty.

Closes #492.

Phyks committed:
Use build status on master branch in README.md (#564)

Use the build status badge from master branch in README.md, to avoid having a "build failed" icon, while the master branch still builds fine.

Phyks committed:
Fix npm dependencies (#563)

* [fix] incompatible dependencies
* [chore] bump Node modules to latest version
* [fix] linting

Closes #500.

Phyks committed:
Fix NPM dependencies (#576)

Commit:
built

Commit:
0.7.3

Phyks committed:
Dev tools fixes (#560)

* Dev tools fixes
* Fix split on `:` in dev tools which was preventing from setting a field
to a URL value.
See https://github.com/cozy-labs/konnectors/issues/547#issue-185478387.
* Use JSON description to set fields through the dev tools, which is a
more robust and reliable way to describe the fields values, without
parsing errors due to commas or colons.
* Fix `konnector_hash` and dev tools to be able to use ES6 konnectors as
well as JS folder import as a package.
* Introduce a new script to build server part only, which can be called
from `npm run build:server`.
* Convert dev tools code to ES6, using
[`decaffeinate`](https://github.com/decaffeinate/decaffeinate).
* Fix the dev tools to look for modules in `build` directory rather than
`server` directory directly. This makes it compatible with
ES6/Coffee/JS, at the expense of having to run a `npm run build:server`
before.
* Add a `npm run dev_tools` endpoint to ease launch of dev tools.
* `vendor` is not a Konnector model attribute and then should not be in
the ouput of `--list` command in the dev tools.
* `--change` should not allow one to set fields not declared in the
connector definition.
Fixes #547.
Fixes #553.
* Update .gitignore to ignore Vagrant files

Phyks committed:
Write a weboob lib
to generate konnectors using Weboob rather than having a single Weboob konnector

Phyks committed:
Convert models and fields in weboob modules export

Phyks committed:
Continue automatic generation of Weboob-based konnectors

Phyks commented:
(rewrote the history to purge built files and ease readability)

Phyks committed:
Continue fetchOperations for generated weboob konnectors

Phyks committed:
Fix missing imports in WeboobFetcher

Phyks committed:
Rewrite parts of Konnectors backend to be asynchronous

This was needed to be able to generate Weboob-backed connectors
automatically and integrate them in the current workflow (conversation
with Cozyweboob is event-based).

Now, when running `npm run dev_tools -- --init`, one gets 252 konnectors
out of the box, most of them being based on Weboob modules. At the
moment, most of them will not be effective as necessary converters have
not yet been implemented and `fetchOperations` have not been fully
tested.

At the moment, I did not took care of the unit tests, but some of them
will have to be fixed.

Note: I am not 100% satisfied by this commit. At least, it makes it
work. But I'd rather go with `Promise` objects, which would be cleaner
in my opinion and avoid a lot of recomputation (anytime `konnector_hash`
is invoked somewhere, it is recomputed at the moment), but needs a much
more important refactoring. :/

Phyks committed:
Add support for folder input

nono commented:
@cozy-labs/konnector-maintainers hi, @Phyks would like to have feedbacks on his work. I'll try to review this PR but you know konnectors better than me. In particular, he had to make functions that were synchronous so far asynchronous (https://github.com/cozy-labs/konnectors/pull/548/commits/0fa603f1e2c378cd0600ac4fc616172aef52e85d).

Phyks committed:
Fix typo in French locale

Phyks commented on .gitmodules:
```diff
@@ -0,0 +1,6 @@
+	url = https://github.com/Phyks/cozyweboob
+[submodule "server/konnectors/weboob/cozyweboob"]
+	path = server/konnectors/weboob/cozyweboob
+	url = https://github.com/Phyks/cozyweboob
```
Indeed. This is a Python code, I should check how I can reliabily distribute it with NPM.

nono commented:
@Phyks hi, I'm sorry. I've said that I will look to this PR but I'm very busy currently and still haven't done it. It's not forgotten, I will try again later.

Phyks commented:
Ok, np. I just wanted some feedback before working further on this, to be sure I was going in the right direction and I will come up with something in agreement with the roadmap for konnectors which will be easily mergeable.

nono commented:
Hi @Phyks, I've read your PR (finally). I don't know konnectors well enough to say what can be improved, but clearly, it goes on the right direction, and it will make sense to use it with the new cozy-stack when konnectors will be ported to it. Thanks for the hard work!

Phyks committed:
Update gitignore to include build folder and avoid adding it by mistake

Phyks committed:
Fix some mistakes

* Remove useless `weboob` konnector, `weboob` will be used as a lib.
* Fix a mistake in dev tools.
* Prevent the UI from spawning multiple Python subprocesses each time
the Konnectors list is to be retrieved.

Phyks committed:
Add a field on Konnector model to indicate whether it comes from Weboob or not
